src/manager.cpp

```diff
@@ -192,20 +198,32 @@ bool Manager::mute_channel(std::string const& name)
     return true;
 }
 
-ContextInfo const& Manager::global_context_info() const
+std::shared_ptr<ContextInfo const> Manager::global_context_info() const
 {
-    return *global_context_info_;
+    return std::atomic_load(&global_context_info_);
 }
 
 void Manager::replace_global_context_info(ContextInfo context_info)
 {
-    global_context_info_ = std::make_shared<ContextInfo const>(std::move(context_info));
+    replace_global_context_info_rvalue(std::move(context_info));
+}
+
+void Manager::replace_global_context_info_rvalue(ContextInfo&& context_info)
+{
+    // First allocate the new ContextInfo, and then atomically replace the pointer held in global_context_info_
+    auto new_context_info = std::make_shared<ContextInfo const>(context_info);
+    std::atomic_store(&global_context_info_, new_context_info);
 }
 
-void Manager::update_global_context_info(ContextInfo context_info)
+// Calling this method concurrently from multiple threads could result in loss of context info (one of the calls could
+// be lost)
+void Manager::update_global_context_info(ContextInfo const& new_context_info)
```

Comment:
Not sure I understand your question.
I fixed here a bug in the implementation, previously updating wouldn't actually update existing kets.

Correct me if I'm wrong, but I don't think the performance of `update_global_context_info`is super-critical, since it's not expected to be called often, only upon specific events which change the context.

See what I wrote in the PR description:
1. bugfix: `update_global_context_info` now updates properly so that overridden keys will actually override

Example:
```
auto old_ci = manager.global_context_info();
ContextInfo ci{{"key1", "value1"}, {"key2", "value2"}, {"key3", "value3"}};
manager.replace_global_context_info(ci);
manager.update_global_context_info({{"key2", "new_value2"}});
```
The above code would result in `"key2"` having `"value2"` instead of being overwritten with `"new_value_2"`.